File: src/agents/strategy-agent.ts
/**
 * Strategy Agent - 融资策略与尽调支持代理
 */

import type {
  SkillContext,
  SkillResponse,
  UserIntent,
  FundraisingStrategy,
  DueDiligenceChecklist,
} from '../types.js';

export class StrategyAgent {
  async initialize(context: SkillContext): Promise<void> {
    console.log('[Strategy Agent] 初始化完成');
  }

  /**
   * 创建融资策略
   */
  async createStrategy(
    intent: UserIntent | Record<string, any>,
    context: SkillContext
  ): Promise<SkillResponse> {
    try {
      const projectInfo = intent.entities?.projectInfo || intent;

      // 使用 LLM 生成融资策略
      const response = await context.llm.chat({
        system: this.getStrategySystemPrompt(),
        messages: [
          {
            role: 'user',
            content: `请为以下项目制定融资策略:\n${JSON.stringify(projectInfo, null, 2)}`,
          },
        ],
      });

      // 解析策略（简化版，实际应该更复杂）
      const strategy = this.parseStrategy(response.content);

      return {
        type: 'text',
        content: this.formatStrategy(strategy),
        metadata: { strategy },
      };
    } catch (error) {
      return {
        type: 'error',
        content: `制定融资策略时出错: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * Term Sheet 解读
   */
  async reviewTermSheet(
    intent: UserIntent | Record<string, any>,
    context: SkillContext
  ): Promise<SkillResponse> {
    try {
      const termSheet = intent.entities?.termSheet || intent.termSheet;

      const response = await context.llm.chat({
        system: this.getTermSheetSystemPrompt(),
        messages: [
          {
            role: 'user',
            content: `请解读以下 Term Sheet 并提供谈判建议:\n${termSheet}`,
          },
        ],
      });

      return {
        type: 'text',
        content: response.content,
      };
    } catch (error) {
      return {
        type: 'error',
        content: `解读 Term Sheet 时出错: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * 准备尽调材料
   */
  async prepareDueDiligence(
    intent: UserIntent | Record<string, any>,
    context: SkillContext
  ): Promise<SkillResponse> {
    try {
      const checklist = this.generateDDChecklist();

      return {
        type: 'text',
        content: this.formatDDChecklist(checklist),
        metadata: { checklist },
      };
    } catch (error) {
      return {
        type: 'error',
        content: `生成尽调清单时出错: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * 行业分析
   */
  async analyzeIndustry(
    intent: UserIntent | Record<string, any>,
    context: SkillContext
  ): Promise<SkillResponse> {
    try {
      const industry = intent.entities?.industry || intent.industry;

      let searchResults = [];
      if (context.tools.searchWeb && context.config.enableWebSearch) {
        // 使用网络搜索获取最新行业数据
        searchResults = await context.tools.searchWeb(
          `${industry} 行业趋势 投资 2024-2026`
        );
      }

      const response = await context.llm.chat({
        system: this.getIndustryAnalysisSystemPrompt(),
        messages: [
          {
            role: 'user',
            content: `请分析 ${industry} 行业的投资趋势、市场规模和竞争格局。\n\n参考资料:\n${searchResults.map(r => `- ${r.title}: ${r.snippet}`).join('\n')}`,
          },
        ],
      });

      return {
        type: 'text',
        content: response.content,
      };
    } catch (error) {
      return {
        type: 'error',
        content: `分析行业时出错: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * System Prompts
   */
  private getStrategySystemPrompt(): string {
    return `你是一位资深的融资顾问，专注于帮助创业公司制定融资策略。

请根据项目信息提供以下内容：
1. 推荐的融资轮次和金额
2. 预计稀释比例
3. Pre-money 估值建议
4. 融资时间线（各阶段及时长）
5. 关键里程碑
6. Pitch 策略要点
7. 谈判技巧

以结构化的方式输出，包含具体的数字和可执行的建议。`;
  }

  private getTermSheetSystemPrompt(): string {
    return `你是一位精通投资条款的法律和财务顾问。

请解读 Term Sheet 中的关键条款，包括但不限于：
- 估值和定价
- 优先清算权
- 反稀释条款
- 董事会席位
- 保护性条款
- 拖售权和跟售权
- 回购权

对每个条款：
1. 解释其含义
2. 分析对创始人的影响
3. 提供谈判建议
4. 指出需要特别注意的风险点`;
  }

  private getIndustryAnalysisSystemPrompt(): string {
    return `你是一位行业分析专家，专注于一级市场投资研究。

请提供全面的行业分析报告，包括：
1. 行业概况和定义
2. 市场规模和增长趋势
3. 产业链结构
4. 竞争格局和主要玩家
5. 投资热度和趋势
6. 关键成功因素
7. 风险和挑战
8. 未来展望

使用数据支撑观点，引用可靠来源。`;
  }

  /**
   * 解析融资策略（简化版）
   */
  private parseStrategy(content: string): FundraisingStrategy {
    // 实际应该用更复杂的解析逻辑
    return {
      recommended_round: 'Pre-A',
      target_amount: 2000,
      dilution_estimate: 15,
      pre_money_valuation: 8000,
      timeline: [
        {
          phase: '准备阶段',
          duration: '4-6 周',
          activities: ['完善 BP', '准备财务数据', '整理尽调材料'],
        },
        {
          phase: '接触阶段',
          duration: '6-8 周',
          activities: ['初步接触投资人', '发送 BP', '安排初步会议'],
        },
        {
          phase: '尽调阶段',
          duration: '4-6 周',
          activities: ['配合尽调', '回答问题', '提供补充材料'],
        },
        {
          phase: '谈判签约',
          duration: '2-4 周',
          activities: ['Term Sheet 谈判', '法律文件审核', '签约交割'],
        },
      ],
      key_milestones: [
        {
          title: 'BP 定稿',
          description: '完成商业计划书并获得 3 位行业专家反馈',
          deadline: '第 2 周',
          importance: 'critical',
        },
        {
          title: '首轮接触',
          description: '与 20 家目标投资机构建立初步联系',
          deadline: '第 8 周',
          importance: 'high',
        },
      ],
      pitch_strategy: [
        '开场 30 秒抓住注意力，突出核心价值主张',
        '用数据说话，展示增长势头',
        '讲故事，让投资人记住你的愿景',
        '提前准备常见问题的答案',
      ],
      negotiation_tips: [
        '不要急于接受第一个 offer',
        '争取更高估值，但要有数据支撑',
        '关注条款的长期影响，不只是估值',
        '保持多个投资人同时推进，增加议价能力',
      ],
    };
  }

  /**
   * 格式化融资策略
   */
  private formatStrategy(strategy: FundraisingStrategy): string {
    let content = `🚀 **融资策略方案**\n\n`;

    content += `**融资概要**\n`;
    content += `- 推荐轮次: ${strategy.recommended_round}\n`;
    content += `- 目标金额: ${strategy.target_amount.toLocaleString()} 万元\n`;
    content += `- 预计稀释: ${strategy.dilution_estimate}%\n`;
    content += `- Pre-money 估值: ${strategy.pre_money_valuation.toLocaleString()} 万元\n\n`;

    content += `**融资时间线** (预计 16-24 周)\n`;
    strategy.timeline.forEach((phase, i) => {
      content += `\n${i + 1}. **${phase.phase}** (${phase.duration})\n`;
      phase.activities.forEach((activity) => {
        content += `   • ${activity}\n`;
      });
    });

    content += `\n**关键里程碑**\n`;
    strategy.key_milestones.forEach((milestone) => {
      const icon =
        milestone.importance === 'critical'
          ? '🔴'
          : milestone.importance === 'high'
          ? '🟡'
          : '🟢';
      content += `\n${icon} **${milestone.title}** (${milestone.deadline})\n`;
      content += `   ${milestone.description}\n`;
    });

    content += `\n📊 **Pitch 策略**\n`;
    strategy.pitch_strategy.forEach((tip) => {
      content += `• ${tip}\n`;
    });

    content += `\n🤝 **谈判建议**\n`;
    strategy.negotiation_tips.forEach((tip) => {
      content += `• ${tip}\n`;
    });

    return content;
  }

  /**
   * 生成尽调清单
   */
  private generateDDChecklist(): DueDiligenceChecklist {
    return {
      categories: [
        {
          name: '公司基本信息',
          items: [
            {
              title: '营业执照',
              description: '公司营业执照副本',
              required: true,
              documents: ['营业执照扫描件', '最新年检记录'],
            },
            {
              title: '公司章程',
              description: '现行有效的公司章程',
              required: true,
              documents: ['公司章程', '历次章程修正案'],
            },
            {
              title: '股权结构',
              description: '完整的股权结构和股东名册',
              required: true,
              documents: ['股东名册', '股权架构图', '代持协议（如有）'],
            },
          ],
        },
        {
          name: '财务信息',
          items: [
            {
              title: '财务报表',
              description: '近三年及最近一期财务报表',
              required: true,
              documents: [
                '资产负债表',
                '利润表',
                '现金流量表',
                '审计报告（如有）',
              ],
            },
            {
              title: '税务文件',
              description: '纳税申报和完税证明',
              required: true,
              documents: ['纳税申报表', '完税证明', '税务合规证明'],
            },
            {
              title: '银行账户',
              description: '公司银行账户信息和流水',
              required: true,
              documents: ['银行开户许可证', '近6个月银行流水'],
            },
          ],
        },
        {
          name: '业务运营',
          items: [
            {
              title: '商业合同',
              description: '重要的商业合同和协议',
              required: true,
              documents: ['客户合同清单', '供应商合同清单', '重大合同原件'],
            },
            {
              title: '知识产权',
              description: '专利、商标、著作权等',
              required: true,
              documents: [
                '专利证书',
                '商标注册证',
                '软件著作权证书',
                'IP 清单',
              ],
            },
            {
              title: '产品技术',
              description: '产品介绍和技术文档',
              required: false,
              documents: ['产品说明书', '技术白皮书', '研发文档'],
            },
          ],
        },
        {
          name: '人力资源',
          items: [
            {
              title: '组织架构',
              description: '公司组织架构和人员名单',
              required: true,
              documents: ['组织架构图', '员工花名册'],
            },
            {
              title: '核心团队',
              description: '创始人和核心团队简历',
              required: true,
              documents: ['创始人简历', '核心团队背景调查'],
            },
            {
              title: '劳动合同',
              description: '员工劳动合同和社保缴纳',
              required: true,
              documents: ['劳动合同样本', '社保缴纳证明'],
            },
          ],
        },
        {
          name: '法律合规',
          items: [
            {
              title: '诉讼仲裁',
              description: '历史和进行中的诉讼、仲裁',
              required: true,
              documents: ['诉讼清单', '法院判决书（如有）'],
            },
            {
              title: '行政处罚',
              description: '行政处罚和监管措施',
              required: true,
              documents: ['行政处罚决定书（如有）', '合规声明'],
            },
            {
              title: '关联交易',
              description: '关联方及关联交易情况',
              required: true,
              documents: ['关联方清单', '关联交易明细'],
            },
          ],
        },
      ],
      estimated_time: '4-6 周',
      tips: [
        '提前准备，避免临时抱佛脚',
        '建立电子数据室，方便投资人查阅',
        '对敏感信息做好脱敏处理',
        '指定专人负责对接尽调事宜',
        '对可能的问题提前准备解释说明',
      ],
    };
  }

  /**
   * 格式化尽调清单
   */
  private formatDDChecklist(checklist: DueDiligenceChecklist): string {
    let content = `📋 **尽职调查材料清单**\n\n`;
    content += `预计准备时间: ${checklist.estimated_time}\n\n`;

    checklist.categories.forEach((category, i) => {
      content += `**${i + 1}. ${category.name}**\n\n`;

      category.items.forEach((item) => {
        const requiredIcon = item.required ? '🔴 [必需]' : '⚪ [可选]';
        content += `${requiredIcon} **${item.title}**\n`;
        content += `   ${item.description}\n`;
        content += `   所需文件:\n`;
        item.documents.forEach((doc) => {
          content += `   • ${doc}\n`;
        });
        content += `\n`;
      });
    });

    content += `💡 **准备建议**\n`;
    checklist.tips.forEach((tip) => {
      content += `• ${tip}\n`;
    });

    return content;
  }
}

File: fa_plugin/agents/strategy_agent.py
"""Strategy Agent - 融资策略与尽调支持代理."""

from typing import Any, Dict

from fa_plugin.agents.base import BaseAgent
from fa_plugin.types.models import UserIntent


class StrategyAgent(BaseAgent):
    """融资策略Agent"""

    async def initialize(self, context: Any) -> None:
        """初始化"""
        self.logger.info("[Strategy Agent] 初始化完成")
        self.initialized = True

    async def handle(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """处理策略相关请求（统一入口）"""
        return await self.create_strategy(intent, context)

    async def create_strategy(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """
        创建融资策略

        Args:
            intent: 用户意图
            context: FA Context

        Returns:
            响应字典
        """
        try:
            project_info = intent.entities.get("project_info", {})

            # 使用LLM生成策略
            strategy = await self._generate_strategy(project_info, context)

            # 格式化输出
            formatted = self._format_strategy(strategy)

            return self._format_response(
                f"🚀 **融资策略方案**\n\n{formatted}",
                strategy=strategy,
            )

        except Exception as e:
            return self._format_error(e)

    async def review_term_sheet(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """Term Sheet解读"""
        try:
            term_sheet_content = intent.entities.get("term_sheet", "")

            # 使用LLM分析
            analysis = await self._analyze_term_sheet(term_sheet_content, context)

            return self._format_response(
                f"📄 **Term Sheet 解读**\n\n{analysis}",
                analysis=analysis,
            )

        except Exception as e:
            return self._format_error(e)

    async def prepare_due_diligence(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """准备尽调材料"""
        try:
            # 生成尽调清单
            checklist = self._generate_dd_checklist()

            formatted = self._format_dd_checklist(checklist)

            return self._format_response(
                f"📋 **尽职调查材料清单**\n\n{formatted}",
                checklist=checklist,
            )

        except Exception as e:
            return self._format_error(e)

    async def analyze_industry(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """行业分析"""
        try:
            industry = intent.entities.get("industry", "")

            # 使用LLM进行行业分析
            analysis = await self._analyze_industry_with_llm(industry, context)

            return self._format_response(
                f"📈 **{industry} 行业分析报告**\n\n{analysis}",
                analysis=analysis,
            )

        except Exception as e:
            return self._format_error(e)

    async def _generate_strategy(
        self, project_info: Dict[str, Any], context: Any
    ) -> Dict[str, Any]:
        """生成融资策略"""
        system_prompt = """你是专业的融资顾问，请为项目制定融资策略。

包括：
1. 推荐融资轮次
2. 目标融资金额
3. 预计稀释比例
4. 融资时间表
5. 关键里程碑
6. 融资策略建议

以JSON格式返回。"""

        response = await context.llm.chat(
            system=system_prompt,
            messages=[{"role": "user", "content": f"项目信息: {project_info}"}],
        )

        # 简化：返回结构化数据
        return {
            "recommended_round": "Pre-A",
            "target_amount": 2000,
            "dilution": 15,
            "pre_money_valuation": 11333,
            "timeline": [
                {"phase": "准备阶段", "duration": "4-6周", "activities": ["完善BP", "准备财务模型"]},
                {"phase": "接触阶段", "duration": "8-12周", "activities": ["接触投资人", "进行路演"]},
            ],
            "milestones": [
                {"title": "完成BP", "deadline": "2周内", "importance": "critical"},
                {"title": "财务模型", "deadline": "3周内", "importance": "high"},
            ],
            "pitch_tips": ["突出核心竞争力", "展示团队优势", "说明市场机会"],
        }

    async def _analyze_term_sheet(self, content: str, context: Any) -> str:
        """分析Term Sheet"""
        system_prompt = """你是专业的投资律师，请分析这份Term Sheet，重点关注：
1. 估值条款
2. 优先权条款
3. 董事会席位
4. 反稀释保护
5. 清算优先权
6. 拖售权/跟售权

指出关键条款和潜在风险。"""

        response = await context.llm.chat(
            system=system_prompt,
            messages=[{"role": "user", "content": content}],
        )

        return response.get("content", "分析结果")

    def _generate_dd_checklist(self) -> Dict[str, Any]:
        """生成尽调清单"""
        return {
            "estimated_time": "4-6周",
            "categories": [
                {
                    "name": "公司基本信息",
                    "items": [
                        {
                            "title": "营业执照",
                            "required": True,
                            "documents": ["营业执照副本", "章程"],
                        },
                        {
                            "title": "股权结构",
                            "required": True,
                            "documents": ["股东名册", "股权架构图"],
                        },
                    ],
                },
                {
                    "name": "财务信息",
                    "items": [
                        {
                            "title": "财务报表",
                            "required": True,
                            "documents": ["近3年审计报告", "最近财务报表"],
                        }
                    ],
                },
            ],
            "tips": ["提前准备材料", "确保数据真实性", "设置数据室"],
        }

    async def _analyze_industry_with_llm(self, industry: str, context: Any) -> str:
        """使用LLM进行行业分析"""
        system_prompt = f"""你是行业分析专家，请分析{industry}行业：

1. 行业概况
2. 市场规模
3. 发展趋势
4. 竞争格局
5. 投资热度
6. 未来展望

请提供专业、详细的分析。"""

        response = await context.llm.chat(
            system=system_prompt,
            messages=[{"role": "user", "content": f"分析{industry}行业"}],
        )

        return response.get("content", "分析结果")

    def _format_strategy(self, strategy: Dict[str, Any]) -> str:
        """格式化策略输出"""
        return f"""
**融资概要**
- 推荐轮次: {strategy['recommended_round']}
- 目标金额: {strategy['target_amount']} 万元
- 预计稀释: {strategy['dilution']}%
- Pre-money估值: {strategy['pre_money_valuation']} 万元

**融资时间表**
{chr(10).join([f"{i+1}. {t['phase']} ({t['duration']})" for i, t in enumerate(strategy.get('timeline', []))])}

**关键里程碑**
{chr(10).join([f"• {m['title']} - {m['deadline']}" for m in strategy.get('milestones', [])])}

**融资策略建议**
{chr(10).join([f"• {tip}" for tip in strategy.get('pitch_tips', [])])}
"""

    def _format_dd_checklist(self, checklist: Dict[str, Any]) -> str:
        """格式化尽调清单"""
        result = [f"**预计准备时间**: {checklist['estimated_time']}\n"]

        for cat in checklist.get("categories", []):
            result.append(f"\n**{cat['name']}**")
            for item in cat["items"]:
                required = "🔴 [必需]" if item["required"] else "🟡 [建议]"
                result.append(f"{required} {item['title']}")
                result.append(f"   所需文件: {', '.join(item['documents'])}")

        result.append("\n**💡 准备建议**")
        for tip in checklist.get("tips", []):
            result.append(f"• {tip}")

        return "\n".join(result)
